### PR TITLE
Feat: update eagle3 example; add export

### PR DESCRIPTION
**Pull request by h-guo18**
## What does this PR do?

**Type of change:** New feature, New example; 

**Overview:** 
Add necessary changes and new features for updating eagle example, including:

Feat: add export support for speculative decoding models in hf_unified_export;
Update several arguments names and default values for example simplicity;
Added a few new scripts and renamed some files for the example;
Rewrote the README:

Rearrange content order; Introduce a "simplified workflow" section;
Provided more details in the "Complete workflow" section.
Removed deprecated contents: Nemo link and notebook example;

## Usage
See README.md for usage.

```python
# Add a code snippet demonstrating how to use this
```

## Testing

Tested dummy training + ar_validate + export with:
- Llama3.2-1B
- Qwen
- Mixtral
- Phi-3

Tested deployment on:
- trtllm-serve
- SGLang

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

- New Features
  - End-to-end train→validate→export workflow with one-line orchestrator, HF checkpoint export supporting speculative-decoding/Eagle format, optional W&B logging, and a new training launch script.

- Documentation
  - README reorganized into simplified and complete workflows, setup/prereqs, SLURM data-prep guide, dataset instructions, and expanded deployment guidance (TRT-LLM, SGLang, quantized).

- Changes
  - AR validation and CLI defaults adjusted (more samples/steps, chat enabled by default), draft-vocab sizing moved to explicit CLI, Eagle config updated (rope-scaling, init range).

- Bug Fixes
  - Centralized draft-vocab mapping, device-alignment fixes, and per-step training accuracy reported in outputs.

### CODE REVIEW

File: examples/speculative_decoding/export_hf_checkpoint.py
Comment (yeyu-nvidia) [inline examples/speculative_decoding/export_hf_checkpoint.py:1]: What is this file for? Can you add some docstring to explain?

File: modelopt/torch/export/unified_export_hf.py [hunk 1]
```diff
@@ -509,12 +521,16 @@ def export_hf_checkpoint(
     try:
         post_state_dict, hf_quant_config = _export_hf_checkpoint(model, dtype)
 
-        # Save hf_quant_config.json for backward compatibility
-        with open(f"{export_dir}/hf_quant_config.json", "w") as file:
-            json.dump(hf_quant_config, file, indent=4)
+        # When there's no quantization applied, we avoid saving hf_quant_config.json for compatibility
```
Comment (yeyu-nvidia) [inline modelopt/torch/export/unified_export_hf.py]: I think we need to revert this unless the team agrees on removing it if no quant applied.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 2]
```diff
@@ -842,10 +852,12 @@ def forward(
 
             if not isinstance(eagle_cache, Cache):
                 eagle_cache = DynamicCache.from_legacy_cache(eagle_cache)
-            past_key_values.eagle_cache = eagle_cache
+
+            # NOTE: diabled for now.
+            # past_key_values.eagle_cache = eagle_cache
```
Comment (yeyu-nvidia) [inline modelopt/torch/speculative/plugins/transformers.py]: This may break model.generate

Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: Can you test if hf_ptq.py still works?

File: examples/speculative_decoding/README.md [hunk 3]
```diff
@@ -1,136 +1,142 @@
...
+[![Documentation](https://img.shields.io/badge/Docs-TensorRT--Model--Optimizer-blue?logo=readthedocs&style=flat-square)](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/5_speculative_decoding.html)
 
-This section focuses on the end-to-end workflow of training speculative decoding modules to deploy for your model.
+Speculative decoding accelerates auto-regressive generation in large language models (LLMs) by leveraging a lightweight draft model to predict the next γ tokens. The main LLM then verifies these candidate tokens in a single forward pass. If the draft model correctly predicts α tokens, the LLM can accept and generate α+1 tokens per verification step, significantly improving throughput.
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]:
```suggestion
Speculative decoding accelerates auto-regressive generation in large language models (LLMs) by leveraging a lightweight draft model to predict the next γ tokens. The main LLM then verifies these candidate tokens in a single forward pass. If the draft model correctly predicts α tokens, the LLM can accept and generate α+1 tokens per verification step, significantly improving generation speed.
```

File: examples/speculative_decoding/README.md [hunk 4]
```diff
@@ -1,136 +1,142 @@
...
-| Support Matrix | View the support matrix to see speculation technique support | \[[Link](#support-matrix)\] | |
-| End to End | Example scripts demonstrating how to train speculation modules using Hugging Face / NeMo / Megatron-LM models | \[[Link](#end-to-end-speculative-decoding-examples)\] | |
-| Deployment | Next steps after speculation module is trained | \[[Link](#deployment)\] | |
-| Speculation Module Checkpoints | View pre-trained speculation modules ready to deploy! | \[[Link](#speculation-module-checkpoints)\] | \[[docs](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/1_quantization.html)\] |
-| Resources | Extra links to relevant resources | \[[Link](#resources)\] | |
+| **Section** | **Description** | **Jump To** |
+| :------------: | :------------: | :------------: |
+| Pre-Requisites | Required & optional dependencies | \[[Link](#pre-requisites)\] |
+| Simplified Workflow | Train, evaluate and export eagle model with one-line command | \[[Link](#getting-started-simplified-workflow)\] |
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]:
```suggestion
| Simplified Workflow | Train, evaluate, and export eagle model with one-line command | \[[Link](#getting-started-simplified-workflow)\] |
```

File: examples/speculative_decoding/README.md [hunk 5]
```diff
@@ -1,136 +1,142 @@
...
 
-```sh
-git clone https://huggingface.co/datasets/nvidia/Daring-Anteater
-python3 server_generate.py --data_path Daring-Anteater/train.jsonl --output_path finetune/data.jsonl --max_token 512 --chat
-```
+- Initializes the draft model with [default settings](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/speculative/eagle/default_config.py#L18)
+- Fine-tunes the model on the [Daring-Anteater](https://huggingface.co/datasets/nvidia/Daring-Anteater) dataset
+- Evaluates the acceptance rate on [MT-Bench](https://huggingface.co/datasets/HuggingFaceH4/mt_bench_prompts)
+- Exports a checkpoint ready for deployment
+- Run an interactive inference demo with vLLM using provided prompt
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]: Does vllm work with the checkpoint? Why not use TRTLLM?

File: examples/speculative_decoding/README.md [hunk 6]
```diff
@@ -1,136 +1,142 @@
...
 
-Create shards of some given size
+Note: Add `--quantization=modelopt` flag for quantized models.
 
-```sh
-python3 distributed_generate/sharding_utils.py --input_path /data/train.jsonl --output_dir /data/train/ --max_lines_per_shard 10000
+Then, we generate conversations with base model and prompts from Daring-Anteater:
+
+```bash
+python server_generate.py --data_path Daring-Anteater/train.jsonl --output_path synthetic/train.jsonl
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md:74]: You removed introduction for --system_prompt. Why? This is important for some benchmark.

Comment (yeyu-nvidia) [reply examples/speculative_decoding/README.md:74]: Also the distributed data generation using SLURM is very helpful when you need large scale data generation. Why do you remove that?

File: examples/speculative_decoding/README.md [hunk 7]
```diff
@@ -1,136 +1,142 @@
...
+
+```bash
+python calibrate_draft_vocab.py --model meta-llama/Llama-3.2-1B-Instruct --data Daring-Anteater/train.jsonl ----draft_vocab_size 32000 --save_dir draft_vocab_cache
 ```
 
-`/scripts/` is the absolute path to `modelopt/examples/speculative_decoding` which contains `server_generate.py` and `distributed_generate`.
-This will launch a vllm server (sglang is also available) on each node. Each node will work through 10 shards of data (10\*max_lines_per_shard number of samples).
-In this case, the first 40 shards of data will be processed.
-To process the next 40 shards
+This will produce a `d2t.pt` file in `save_dir`, which is the mapping from draft vocabs to full vocab that will be read by our draft model later.
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]: Better to explain how this mapping works. i.e., tokens = draft_tokens + d2t[draft_tokens]

File: examples/speculative_decoding/README.md [hunk 8]
```diff
@@ -1,136 +1,142 @@
...
 
-```sh
-bash distributed_generate/launch.sh $SLURM_JOB_ID vllm TinyLlama/TinyLlama-1.1B-Chat-v1.0 /data/train/ /data/output /scripts/ 40 10 n1,n2,n3,n4
-```
+### (Optional) Configuring Draft Model
 
-To combine the shards back
+For eagle1 and eagle3 we provide an [default model architecture config](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/speculative/eagle/default_config.py#L18) in modelopt. User can overwrite default settings by providing additional json dict. In this example, we overwrite the `draft_vocab_size` by in `eagle_config.json`:
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]: this default config sets "use_aux_hidden_state" to False so it's just eagle1. it is modified to eagle3 in https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/speculative/config.py#L27

File: examples/speculative_decoding/README.md [hunk 9]
```diff
@@ -1,136 +1,142 @@
...
+{
+    "draft_vocab_size": 32000
+}
 ```
 
-### Speculative Decoding Example Training Workflow
+### Training Draft Model with Modelopt
 
-Here is the recommended end-to-end speculative decoding training workflow:
+`main.py` provides a example for converting a base HF model for speculative decoding and training it. It consists of a few simple steps:
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]:
```suggestion
`main.py` provides an example for converting a HF base model for speculative decoding and training it. It consists of a few simple steps:
```

File: examples/speculative_decoding/export_hf_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    parser.add_argument("--export_path", type=str, default="")
+    return parser.parse_args()
+
+
+mto.enable_huggingface_checkpointing()
+
+args = parse_args()
+model = AutoModelForCausalLM.from_pretrained(args.model_path, torch_dtype="auto")
+with torch.inference_mode():
+    export_hf_checkpoint(
```
Comment (sugunav14) [inline examples/speculative_decoding/export_hf_checkpoint.py:44]: Does this need to be in a separate file? Could we just directly call export_hf_checkpoint() wherever its needed?

File: examples/speculative_decoding/README.md [hunk 11]
```diff
@@ -143,81 +149,87 @@ trainer.save_state()
...
+We omitted details like tokenizer initialization for simplicity. A complete training example is provided in `main.py`, along with a bash script to launch the training with huggingface accelrate in `launch_train.sh`, which can be runned by:
 
-### Supported Models/Techniques
+```bash
+./launch_train.sh --model $BASE_MODEL \
+            --output_dir $OUTPUT_DIR \
+            --data $DATA \
+            --num_gpu $NUM_GPU \
+            --num_epochs 10 \
+            --eagle_config eagle_config.json #This is where we overwrite default eagle configs
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]: And optional

File: examples/speculative_decoding/README.md [hunk 12]
```diff
@@ -143,81 +149,87 @@ trainer.save_state()
...
+```
 
-<!-- ### NeMo:
+This will export the model from a modelopt checkpoint to a deployment-compatible formart.
 
-[This NeMo-Run example](../nemo_run/specdec_sft) performs a start-to-finish Speculative-Decoding and SFT process. See the README inside for instructions. -->
+### Deployment
 
-### HuggingFace
+The exported checkpoint can be deployed on TRT-LLM or vLLM.
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md]: Does it work with vllm?

File: examples/speculative_decoding/README.md [hunk 13]
```diff
@@ -143,81 +149,87 @@ trainer.save_state()
...
 
-### Deployment
+Please refer to [vLLM Doc: Speculative Decoding](https://docs.vllm.ai/en/v0.9.0/features/spec_decode.html) for detailed usage.
+
+#### Deploying Quantized model
 
-The final model after end-to-end speculative decoding fine-tuning is similar in architecture to HF models. It can be further optimized through **ModelOpt**, e.g., PTQ and QAT. It can be deployed to TensorRT-LLM (TRTLLM) or to TensorRT just like a regular **ModelOpt** model. See more details on deployment of quantized model to TRTLLM [here](../llm_ptq/README.md).
+See more details on deployment of quantized model to TRTLLM [here](../llm_ptq/README.md).
+
+## Support Matrix
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/README.md:227]: I think we now can support more models. We will need to update this matrix.

File: examples/speculative_decoding/README.md (hunk 5 above)
Comment (h-guo18) [reply examples/speculative_decoding/README.md]: TRTLLM requires docker or some manual set up. Would be a little complicated to put in this simplified workflow...

File: examples/speculative_decoding/README.md (hunk 12 above)
Comment (h-guo18) [reply examples/speculative_decoding/README.md]: Without the empty `quant_config.json` it works fine with vLLM.

File: examples/speculative_decoding/README.md (hunk 13 above)
Comment (h-guo18) [reply examples/speculative_decoding/README.md:227]: OK. Can you share the new model list? I will need to test them on this HF training code path before updating in README.

File: examples/speculative_decoding/export_hf_checkpoint.py (hunk 10 above)
Comment (h-guo18) [reply examples/speculative_decoding/export_hf_checkpoint.py:44]: There are a few minor cases where we need this:
- When we store/load and continue training but only want to export the final ckpt from last run;
- When we save multiple intermediate checkpoints, evaluate them, but only export one of them with best accuracy;
- When we apply other model optimizations like quantization after mtsp, and does not need to export before that;

We can certainly bake in the above export logics in training scripts, but that would add a little complexity to the example. I think keeping it in a small module would make things concise.

Let me know what do you think. Thanks!

File: examples/speculative_decoding/README.md (hunk 6 above)
Comment (h-guo18) [reply examples/speculative_decoding/README.md:74]: I see. Will put back the content. I think for conciseness it might be a good idea to put slurm section into a separate md file, since it's quite long and not all users need it. (Especially on current supported small models).

**Comment by ChenhanYu** [comment]
https://gitlab-master.nvidia.com/omniml/modelopt/-/pipelines/34540942

File: examples/speculative_decoding/README.md (hunk 13 above)
Comment (yeyu-nvidia) [reply examples/speculative_decoding/README.md:227]: I dont have a list. Theoretically we should be able to support any HF models. Maybe just delete this section?

**Comment by yeyu-nvidia** [comment]
This needs to be addressed https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/293#discussion_r2325925837. Otherwise, LGTM.

File: modelopt/torch/export/unified_export_hf.py [hunk 14]
```diff
@@ -490,6 +491,12 @@ def _export_hf_checkpoint(
     return quantized_state_dict, quant_config
 
 
+def _quant_applied(hf_quant_config: dict) -> bool:
```
Comment (sugunav14) [inline modelopt/torch/export/unified_export_hf.py]: do we need this? I don't see it being used anywhere

Comment (h-guo18) [reply modelopt/torch/export/unified_export_hf.py]: You are right. This is not needed now. Removed now!

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Modify state_dict and config for exporting speculative decoding in official format."""
+
+import torch
+import torch.nn as nn
+
+EAGLE_MODELOPT_TO_OFFICIAL = {
+    "required": {
```
Comment (cjluo-nv) [inline modelopt/torch/export/plugins/hf_spec_export.py:22]: is this from megatron to HF?

Comment (h-guo18) [reply modelopt/torch/export/plugins/hf_spec_export.py:22]: This is the param name mapping from [modelopt::HFEagleModel](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/6b2411e254a432f6d7fd97a664dbf4bb3df0bd5c/modelopt/torch/speculative/plugins/transformers.py#L333) to [eagle3 official checkpoint](https://huggingface.co/yuhuili/EAGLE-LLaMA3-Instruct-8B). I bielieve both megatron and HF export to the same format.

Comment (h-guo18) [reply modelopt/torch/export/plugins/hf_spec_export.py:22]: The latter is also consistent with our [previous released eagle checkpoints](https://huggingface.co/collections/nvidia/model-optimizer-66aa84f7966b3150262481a4).

File: modelopt/torch/export/unified_export_hf.py [hunk 16]
```diff
@@ -509,12 +510,15 @@ def export_hf_checkpoint(
...
         post_state_dict, hf_quant_config = _export_hf_checkpoint(model, dtype)
 
+        # NOTE: (hg) Should we save hf_quant_config when there's no quantization applied?
         # Save hf_quant_config.json for backward compatibility
         with open(f"{export_dir}/hf_quant_config.json", "w") as file:
             json.dump(hf_quant_config, file, indent=4)
 
         hf_quant_config = convert_hf_quant_config_format(hf_quant_config)
 
+        post_state_dict = rename_and_prune_if_spec_decoding(model, post_state_dict)
```
Comment (Edwardf0t1) [inline modelopt/torch/export/unified_export_hf.py:520]: @h-guo18 Do you think we can add an option to add lm-head to the exported eagle ckpt for SGLang deployment? cc @yeyu-nvidia

Comment (h-guo18) [reply modelopt/torch/export/unified_export_hf.py:520]: At the moment, the exported checkpoint always includes the `lm-head`: [link](https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/293/files#diff-5dcaa628649af7aad9a38d598db0db4d78b4ec43516f487baaae2426c3714704R78). 
I was wondering if SGLang expects a different format?
I merged the current version to meet the blog post deadline, but I’d be happy to revise it in a follow-up PR for SGLang compatibility. Thanks!

Comment (Edwardf0t1) [reply modelopt/torch/export/unified_export_hf.py:520]: I see, previously SGLang doesn't support our Eagle3 but after [this PR](https://github.com/sgl-project/sglang/pull/9739) it's all good.
